openverse_catalog/dags/data_refresh/data_refresh_task_factory.py

```diff
@@ -138,37 +140,77 @@ def create_data_refresh_task_group(
+                task_id=task_id,
+                http_conn_id="data_refresh",
+                endpoint="task",
+                method="POST",
+                headers={"Content-Type": "application/json"},
+                data=json.dumps(post_data),
+                response_check=lambda response: response.status_code == 202,
+                response_filter=response_filter_data_refresh,
+            )
+
+        def _get_http_sensor(task_id: str, endpoint: str) -> HttpSensor:
+            """
+            Get an ``HttpSensor`` instance that waits for a given  ingestion server task
+            to be completed. The trigger task status can be observed by polling an
+            endpoint.
+
+            :param task_id: the name of the task associated with the sensor
+            :endpoint: the REST endpoint for tracking the status of the triggered task
+            """
+
+            return HttpSensor(
+                task_id=task_id,
+                http_conn_id="data_refresh",
+                endpoint=endpoint,
+                method="GET",
+                response_check=response_check_wait_for_completion,
+                mode="reschedule",
+                poke_interval=poke_interval,
+                timeout=data_refresh.data_refresh_timeout,
+            )
+
+        tasks = [wait_for_data_refresh]
+        action_data_map: dict[str, dict] = {
+            "ingest_upstream": {},
+            "point_alias": {"alias": data_refresh.media_type},
         }
-
-        # Trigger the refresh on the data refresh server.
-        trigger_data_refresh = SimpleHttpOperator(
-            task_id="trigger_data_refresh",
-            http_conn_id="data_refresh",
-            endpoint="task",
-            method="POST",
-            headers={"Content-Type": "application/json"},
-            data=json.dumps(data_refresh_post_data),
-            response_check=lambda response: response.status_code == 202,
-            response_filter=response_filter_data_refresh,
-        )
-
-        # Wait for the data refresh to complete.
-        wait_for_completion = HttpSensor(
-            task_id="wait_for_completion",
-            http_conn_id="data_refresh",
-            endpoint=XCOM_PULL_TEMPLATE.format(
-                trigger_data_refresh.task_id, "return_value"
-            ),
-            method="GET",
-            response_check=response_check_wait_for_completion,
-            mode="reschedule",
-            poke_interval=poke_interval,
-            timeout=data_refresh.data_refresh_timeout,
-        )
-
-        wait_for_data_refresh >> trigger_data_refresh >> wait_for_completion
+        for action, action_post_data in action_data_map.items():
+            trigger = _get_http_operator(
+                task_id=f"trigger_{action}",
+                post_data=action_post_data
+                | {
+                    "model": data_refresh.media_type,
+                    "action": action.upper(),
+                    "index_suffix": index_suffix,
```

Comment:
I was able to verify this odd behaviour locally. It does not occur when the `index_suffix` is replaced with a hardcoded string so I'm assuming that the UUID is being generated afresh for both tasks in the group (which is strange because it's defined outside the loop).

Reply:
Using a Python operator to set the variable as an XCom (commit ee5b81a) fixes it.

Reply:
FWIW that exactly what is happening! It's worth noting that each task (defined by an operator) is run as an entirely separate Python process. Anything that's defined on-the-fly can be used there, but if it's not _deterministic_ over repeated runs, then you'll get new values. We try to avoid defining those values in the DAG definition script, since they'll be initialized **both** when the DAG is parsed for scheduling and when individual tasks are run. For dynamic values, XComs are indeed the best way to pass data from one task to another in a run (as you've identified!). One exception might be Airflow Variables. We often pull those in within the DAG definition script and hand them to the tasks. They're still generally deterministic though, because we're unlikely to change them mid-DAG-run, so repeated runs will use whatever value the Variable has at that time.

Hope that helps! It's a bit tricky cause it's all in Python, so it's tempting to just handle values like you would with any other Python module 😅